python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py

```diff
@@ -0,0 +1,536 @@
+        errors += self._terminate_errors_queue
+        self._launch_errors_queue = []
+        self._terminate_errors_queue = []
+        return errors
+
+    ############################
+    # Private
+    ############################
+
+    def _initialize_scale_request(
+        self, to_launch: Dict[NodeType, int], to_delete_instances: List[CloudInstanceId]
+    ) -> "KuberayProvider.ScaleRequest":
+        """
+        Initialize the scale request based on the current state of the cluster and
+        the desired state (to launch, to delete).
+
+        Args:
+            to_launch: The desired number of workers to launch for each node type.
+            to_delete_instances: The instances to delete.
+
+        Returns:
+            The scale request.
+        """
+
+        # Update the cached states.
+        self._sync_with_api_server()
+        ray_cluster = self.ray_cluster
+        cur_instances = self.instances
+
+        # Get the worker groups that have pending deletes and the worker groups that
+        # have finished deletes.
+        (
+            worker_groups_with_pending_deletes,
+            worker_groups_without_pending_deletes,
+        ) = self._get_workers_groups_with_deletes(
+            ray_cluster, set(cur_instances.keys())
+        )
+
+        # Calculate the desired number of workers by type.
+        num_workers_dict = defaultdict(int)
+        for instance_id, instance in cur_instances.items():
+            if instance.node_kind == NodeKind.HEAD:
+                # Only track workers.
+                continue
+            num_workers_dict[instance.node_type] += 1
+
+        # Add to launch nodes..
+        for node_type, count in to_launch.items():
+            num_workers_dict[node_type] += count
+
+        # Remove deleting workers.
+        for to_delete_id in to_delete_instances:
+            to_delete_instance = cur_instances.get(to_delete_id, None)
+            if to_delete_instance is None:
+                # This instance has already been deleted.
+                continue
+            num_workers_dict[to_delete_instance.node_type] -= 1
+            assert num_workers_dict[to_delete_instance.node_type] >= 0
+
+        # Group the to delete workers by node type.
+        to_delete_instances_by_type = defaultdict(list)
+        for instance_id in to_delete_instances:
+            instance = cur_instances.get(instance_id, None)
+            if instance is None:
+                # This instance has already been deleted.
+                continue
+
+            if instance.node_kind == NodeKind.HEAD:
+                # Not possible to delete head node.
+                continue
+
+            to_delete_instances_by_type[instance.node_type].append(instance)
+
+        scale_request = KuberayProvider.ScaleRequest(
+            desired_num_workers=num_workers_dict,
+            workers_to_delete=to_delete_instances_by_type,
+            worker_groups_to_clear_delete=worker_groups_without_pending_deletes,
```

Comment:
Can we rename `worker_groups_to_clear_delete` to `worker_groups_without_pending_deletes`?